### PR TITLE
Lvs adk view

### PR DESCRIPTION
**Pull request by steveri**
Simple fix to solve issue https://github.com/StanfordAHA/garnet/issues/752.
Should be pretty self-explanatory when you look at the file-compares.

Proof that it works should be provided by the run I just started...
https://buildkite.com/tapeout-aha/fullchip/builds/205
...which is expected to pass LVS sometime tomorrow ish, as a result of this fix...

### CODE REVIEW

File: steps/mentor-calibre-lvs/configure.yml [hunk 1]
```diff
@@ -71,6 +71,8 @@ parameters:
   #
   lvs_verify_netlist: 1
 
+  lvs_adk: "inputs/adk"
+
```
Comment (ctorng) [inline steps/mentor-calibre-lvs/configure.yml:75]: This isn't how I would approach this change.

Both "inputs" and "adk" are special to mflowgen. The "inputs" directory is one of the few restrictions that mflowgen places on structure, so we need that. The "adk" is actually listed in the inputs of this node, so we cannot change that either.

See below for my recommendation...

File: steps/mentor-calibre-lvs/lvs.runset.template [hunk 2]
```diff
@@ -27,5 +27,5 @@
 *cmnTranscriptEchoToFile: 1
 *cmnRunMT: 1
 *cmnRunHyper: 1
-*cmnV2LVS_SPICEIncFiles: inputs/adk/*.cdl inputs/*.spi inputs/*.sp
+*cmnV2LVS_SPICEIncFiles: ${lvs_adk}/*.cdl inputs/*.spi inputs/*.sp
```
Comment (ctorng) [inline steps/mentor-calibre-lvs/lvs.runset.template:30]: What you actually want is to supply _another_ adk right? One with multi VTs?

So one proposal is:

1. Add an additional parameter for ${lvs_extra_spice_include} to this line here.
2. Attach the 2nd adk as an input
3. Graph should set the parameter "lvs_extra_spice_include" to the 2nd adk's *.cdl

**Review by ctorng — CHANGES_REQUESTED** [review]

File: steps/mentor-calibre-lvs/lvs.runset.template (hunk 2 above)
Comment (steveri) [reply steps/mentor-calibre-lvs/lvs.runset.template:30]: So something like this...? In adks/tsmc16/configure.yml (tsmc16 repo) I add new outputs and such something like
```
outputs:
  - adk_lvs2

commands:
  - 'if [ -z "{adk_lvs2_view}" ];
     then ln -sf /dev/null adk_lvs2;
     else ln -sf ../{adk_lvs2_view} adk_lvs2;
     fi'

parameters:
  adk_lvs2_view: ""
```

Then in full_chip/construct.py (garnet repo) I add new parameters
```
  parameters = {
    # LVS
    'adk_lvs2_view' : 'multivt',
    'lvs_extra_spice_include'   : 'inputs/adk_lvs2/*.cdl',
```

In the lvs step configure.yml (mflowgen repo) I add new inputs and parameters
```
inputs:
  - adk
  - adk_lvs2

parameters:
  lvs_extra_spice_include : ""
```

..and then finally in the lvs template (mflowgen repo) I do this:
```
- *cmnV2LVS_SPICEIncFiles: inputs/adk/*.cdl inputs/*.spi inputs/*.sp
+ *cmnV2LVS_SPICEIncFiles: inputs/adk/*.cdl ${lvs_extra_spice_include} inputs/*.spi inputs/*.sp
```

**Comment by steveri** [comment]
Okay I made the changes (see updated plan above) and am about to start testing...

File: steps/mentor-calibre-lvs/lvs.runset.template [hunk 3]
```diff
@@ -27,5 +27,5 @@
 *cmnTranscriptEchoToFile: 1
 *cmnRunMT: 1
 *cmnRunHyper: 1
-*cmnV2LVS_SPICEIncFiles: inputs/adk/*.cdl inputs/*.spi inputs/*.sp
+*cmnV2LVS_SPICEIncFiles: inputs/adk/*.cdl ${lvs_extra_spice_include} inputs/*.spi inputs/*.sp
```
Comment (ctorng) [inline steps/mentor-calibre-lvs/lvs.runset.template:30]: This part is good.

File: steps/mentor-calibre-lvs/configure.yml [hunk 4]
```diff
@@ -71,6 +78,9 @@ parameters:
   #
   lvs_verify_netlist: 1
 
+  adk_view_lvs2: ""
+  lvs_extra_spice_include: ""
```
Comment (ctorng) [inline steps/mentor-calibre-lvs/configure.yml:82]: We should only keep the "lvs_extra_spice_include" parameter here.

We should set that parameter in our graph directly (with .set_param) -- https://github.com/StanfordAHA/garnet/blob/master/mflowgen/Tile_PE/construct.py#L141

File: steps/mentor-calibre-lvs/configure.yml [hunk 5]
```diff
@@ -29,6 +33,9 @@ outputs:
 commands:
   - if [ -z "$lvs_hcells_file" ]; then export lvs_use_hcells=0; else export lvs_use_hcells=1; fi
   - if [ -z "$lvs_connect_names" ]; then export lvs_connect_names_state=NONE; else export lvs_connect_names_state=SOME; fi
+  - if [ ! -z "$adk_view_lvs2" ];
+    then export lvs_extra_spice_include="inputs/adk_lvs2/*.cdl" ];
+    fi
```
Comment (ctorng) [inline steps/mentor-calibre-lvs/configure.yml:38]: With the comment below (setting it in our graph), we don't need this change.

File: steps/mentor-calibre-lvs/configure.yml [hunk 6]
```diff
@@ -13,6 +16,7 @@ name: mentor-calibre-lvs
 
 inputs:
   - adk
+  - adk_lvs2
```
Comment (ctorng) [inline steps/mentor-calibre-lvs/configure.yml:19]: This is AHA-specific and should be done dynamically in our graph, as opposed to statically baked into the library node.

So use .extend_inputs() like this -- https://github.com/StanfordAHA/garnet/blob/master/mflowgen/Tile_PE/construct.py#L123

File: steps/mentor-calibre-lvs/configure.yml [hunk 7]
```diff
@@ -4,6 +4,9 @@
 # Author : Christopher Torng
 # Date   : November 5, 2019
 #
+# [steveri 01/2021] Added lvs2, lvs_extra_spice etc. to support where e.g.
+# most of the chip uses stdview but one IP block (dragonphy in our case)
+# needs a different view b/c it uses LVT cells etc.
```
Comment (ctorng) [inline steps/mentor-calibre-lvs/configure.yml:9]: So in the end I think the only change is the single line to add the parameter for "lvs_extra_spice_include".

**Review by ctorng — CHANGES_REQUESTED** [review]

**Comment by ctorng** [comment]
Does that make sense?

1. Only have a new line for "lvs_extra_spice_include" in the library node
2. In our graphs...
    a. Plop down a new node for the different adk
    b. Take the lvs step and call .extend_inputs() and give it a new input (for adk2)
    c. Take the lvs step and call .set_param() to set "lvs_extra_spice_include" to "inputs/adk2/*.cdl"
    d. Call g.connect( adk2.o('adk'), lvs.i('adk2') ) to hook up the new adk and the lvs step's adk2... we have to use the .connect() and .i() and .o() instead of .connect_by_name, because the output of your adk2 is "adk" is different from the input of lvs which is "adk2". So you cannot connect by name. Here is an example of connecting differently named inputs/outputs -- https://github.com/StanfordAHA/garnet/blob/master/mflowgen/Tile_PE/construct.py#L256

File: steps/mentor-calibre-lvs/configure.yml (hunk 6 above)
Comment (steveri) [reply steps/mentor-calibre-lvs/configure.yml:19]: > This is AHA-specific and should be done dynamically in our graph, as opposed to statically baked into the library node.
> 
> So use .extend_inputs() like this -- https://github.com/StanfordAHA/garnet/blob/master/mflowgen/Tile_PE/construct.py#L123

RIGHT! extend_inputs(). I forgot about extend_inputs(). That's exactly what I was wanting. Thanks!

**Comment by steveri** [comment]
>     1. Only have a new line for "lvs_extra_spice_include" in the library node
Yes. I like it!

>     2. In our graphs...
>        a. Plop down a new node for the different adk
>        b. Take the lvs step and call .extend_inputs() and give it a new input (for adk2)
>        c. Take the lvs step and call .set_param() to set "lvs_extra_spice_include" to "inputs/adk2/*.cdl"
>        d. Call g.connect( adk2.o('adk'), lvs.i('adk2') ) to hook up the new adk and the lvs step's adk2... we have to use the .connect() and .i() and .o() instead of .connect_by_name, because the output of your adk2 is "adk", which is different from the input of lvs which is "adk2". So you cannot connect by name. Here is an example of connecting differently named inputs/outputs -- https://github.com/StanfordAHA/garnet/blob/master/mflowgen/Tile_PE/construct.py#L256

Hm I think I like my solution better; instead of building a new node, we build two edges coming out of the tsmc16 node, "adk" and "adk_lvs"
```
# a. Add a new edge to existing adk node ("tsmc16")
% ls -l 12-tsmc16/outputs
    adk -> ../stdview/
    adk_lvs2 -> ../multivt/
    pkgs -> ../pkgs/
```
Now we can join the outputs and inputs by name, plus the name actually means something in terms of understanding what the edge is for
```
# b. Use extend_inputs() to connect the edge to lvs
% ls -l 34-mentor-calibre-lvs/inputs/adk*
    adk -> ../../12-tsmc16/outputs/adk/
    adk_lvs -> ../../12-tsmc16/outputs/adk_lvs2
```

**Review by ctorng — APPROVED** [review]

**Pull request merged**